Pass settings as key value or config for storage with settings

kssenii opened the pull request:
Changelog category (leave one):
- Improvement


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
If storage supports SETTINGS allow to pass them as key value or via config. Add this support for mysql.

kssenii committed:
Settings changes as key value

KochetovNicolai reviewed (commented):
Generally looks ok, but could you add some tests?
As I understand, you now we can specify storage settings in config and AST both. I hope those from AST have more priority?

kssenii commented:
> As I understand, you now we can specify storage settings in config and AST both.

Yes, but It is not really about storage settings, it is more about table functions. Some storages have storage settings, but it was never possible to use them for table functions. So now it will be possible to add them either via config or via key-value argument in ast. 

> I hope those from AST have more priority?

Yes, because I in code I firstly check config and then check key-value arguments and override config values by key-value.

> Generally looks ok, but could you add some tests?

I added one test already, here it is:
https://github.com/ClickHouse/ClickHouse/blob/55dfaef4decd8da8b7bf0efc33e90dec82f229f5/tests/integration/test_storage_mysql/test.py#L420-L422
In this test I added a mysql settings as key-value argument: `connection_pool_size=0` and it checks that the setting is used because an exception is expected for value 0 (by default it is non-zero).

Though I haven't added a test for getting setting from config, I will add...

kssenii committed:
Merge remote-tracking branch 'upstream/master' into settings-changes-with-named-conf

kssenii committed:
Add support for dictionary source

kssenii committed:
Add one more test

KochetovNicolai commented on tests/integration/test_storage_mysql/test.py:
```diff
@@ -418,6 +418,9 @@ def test_predefined_connection_configuration(started_cluster):
     assert (node1.query(f"SELECT count() FROM test_table").rstrip() == '100')
 
+    assert 'Connection pool cannot have zero size' in node1.query_and_get_error(f"SELECT count() FROM mysql(mysql1, table='test_table', connection_pool_size=0)")
+    assert 'Connection pool cannot have zero size' in node1.query_and_get_error(f"SELECT count() FROM mysql(mysql4)")
```
Maybe also check that if we add `connection_pool_size=1` it overrides setting from config and works.

kssenii committed:
More tests
